File: Bhumika Agrawal/Stack/Valid Parenthesis.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack=new Stack<>();
        for(char str:s.toCharArray()){
            if(str=='('){
                stack.push(')');
            }
            else if(str=='{'){
                stack.push('}');
            }
            else if(str=='['){
                stack.push(']');
            }
            else{
                if(stack.isEmpty() || stack.peek()!=str){
                    return false;
                }
                else{
                    stack.pop();
                }
            }
        }
        return stack.isEmpty();
        
    }
}
//Time=O(n)
//Space=O(n)

File: Divyansh Singh/Stack/valid parenthesis.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char>st;
        for(int i=0;i<s.length();i++){
            if(s[i]=='('){
               st.push(')');
            }
            else if(s[i]=='{'){
               st.push('}');
            }
            else if(s[i]=='['){
               st.push(']');
            }
            else{
                if(st.empty()==true || st.top()!=s[i]){
                    return false;
                }
                st.pop();
            }
        }

        if(st.empty()){
            return true;
        }
        return false;
    }
};
